File: 93.复原ip地址.go
import (
	"fmt"
	"strconv"
)

/*
 * @lc app=leetcode.cn id=93 lang=golang
 *
 * [93] 复原IP地址
 */

// @lc code=start

const SEG_COUNT = 4

var (
	ans      []string
	segments []int
)

func restoreIpAddresses(s string) []string {
	if s == "" {
		return []string{}
	}

	segments = make([]int, SEG_COUNT)
	ans = []string{}
	dfs(s, 0, 0)
	return ans
}

// 处理一个块，从 segStart 到 s 结束。
func dfs(s string, segIndex, segStart int) {
	if segIndex == SEG_COUNT {
		if segStart == len(s) {
			ipAddr := strconv.Itoa(segments[0])
			for i := 1; i < SEG_COUNT; i++ {
				ipAddr += fmt.Sprintf(".%d", segments[i])
			}
			ans = append(ans, ipAddr)
		}
		return
	}

	if segStart == len(s) {
		return
	}

	// 如果以 0 开头，那这个区块就只能是“0”了。
	if s[segStart] == '0' {
		segments[segIndex] = 0
		dfs(s, segIndex+1, segStart+1)
	}

	addr := 0
	for segEnd := segStart; segEnd < len(s); segEnd++ {
		addr = addr*10 + int(s[segEnd]-'0')
		if addr > 0 && addr <= 255 {
			segments[segIndex] = addr
			dfs(s, segIndex+1, segEnd+1)
		} else {
			break
		}
	}
}

File: 93.复原-ip-地址.java
import java.util.ArrayList;
import java.util.List;

import jdk.javadoc.internal.doclets.formats.html.resources.standard;

/*
 * @lc app=leetcode.cn id=93 lang=java
 *
 * [93] 复原 IP 地址
 */

// @lc code=start
class Solution {
    private List<String> ans;
    private int[] segments;
    private final int SEG_COUNT = 4;

    public List<String> restoreIpAddresses(String s) {
        if (s.length() < 4) {
            return new ArrayList<>();
        }

        segments = new int[SEG_COUNT];
        ans = new ArrayList<>();
        dfs(s, 0, 0);
        return ans;
    }

    private void dfs(String s, int segIndex, int segStart) {
        if (segIndex == SEG_COUNT) {
            if (segStart == s.length()) {
                String ipAddr = segments[0] + "";
                for (int i = 1; i < SEG_COUNT; i++) {
                    ipAddr += ("." + segments[i]);
                }
                ans.add(ipAddr);
            }
            return;
        }

        if (segStart == s.length()) {
            return;
        }

        if (s.charAt(segStart) == '0') {
            segments[segIndex] = 0;
            dfs(s, segIndex + 1, segStart + 1);
        }

        // 表示当前这一部分的地址
        int addr = 0;
        for (int segEnd = segStart; segEnd < s.length(); segEnd++) {
            addr = addr * 10 + (s.charAt(segEnd) - '0');
            if (addr > 0 && addr <= 255) {
                segments[segIndex] = addr;
                dfs(s, segIndex + 1, segEnd + 1);
            } else {
                break;
            }
        }
    }
}
// @lc code=end

